controllers/onload_controller.go

```diff
@@ -432,6 +479,70 @@ func (r *OnloadReconciler) handleModuleUpdate(ctx context.Context, onload *onloa
 	return &ctrl.Result{Requeue: true}, nil
 }
 
+func (r *OnloadReconciler) handleModuleUpdate(ctx context.Context, onload *onloadv1alpha1.Onload) (*ctrl.Result, error) {
+	log := log.FromContext(ctx)
+
+	moduleConfigs := []struct {
+		isUsed     bool
+		moduleName string
+
+		getKernelMapping getKernelMapping
+	}{
+		{
+			true, // Onload module kind is always used for Onload CR
+			onload.Name + onloadModuleNameSuffix,
+			getOnloadKernelMapping,
+		},
+		{
+			onloadUsesSFC(onload),
+			onload.Name + sfcModuleNameSuffix,
+			getSFCKernelMapping,
+		},
+	}
+
+	for _, moduleConfig := range moduleConfigs {
+		module := &kmm.Module{}
+		err := r.Get(ctx, types.NamespacedName{Name: moduleConfig.moduleName, Namespace: onload.Namespace}, module)
+
+		//
+		// Below are possible module kind-related actions.
+		//
+		// +------------------------+----------+----------+
+		// | Exists \ Should exist? |   Yes    |    No    |
+		// +------------------------+----------+----------+
+		// | Yes                    | Patch    | Delete   |
+		// | No                     | Requeue* | Continue |
+		// +------------------------+----------+----------+
+		//
+		// (*) Requeue to let the reconciliation create the new module kind.
+		//
+
+		if err == nil { // Module kind exists
+			if moduleConfig.isUsed {
+				res, err := r.patchModule(ctx, module, onload, moduleConfig.getKernelMapping)
+				if res != nil || err != nil {
+					return res, err
+				}
+			} else {
+				err = r.Delete(ctx, module)
+				return &ctrl.Result{Requeue: true}, err
+			}
+		} else if apierrors.IsNotFound(err) {
+			if moduleConfig.isUsed {
+				return &ctrl.Result{Requeue: true}, nil
```

Comment:
I don't think this case can happen, the reconciliation loop will check the existence of the SFC module before it reaches the upgrade code. Which I think means we can continue in either case if the module doesn't exist.

Reply:
Yes, well spotted. How about raising an error instead because this situation should not occur, as you outline? I believe we'll re-enter reconciliation after that, so it's still not the same as `continue`.